toolchain/check/testdata/impl/no_prelude/impl_cycle.carbon

```diff
@@ -194,3 +194,65 @@ fn F() {
   Compare({} as C, {} as D);
   Compare({} as D, {} as C);
 }
+
+// --- deduce_cycle_without_symbolic.carbon
+library "[[@TEST_NAME]]";
+
+interface First {}
+interface Second {}
+
+// This impl is looked at for "does T impl Second" first, which requires a
```

Comment:
I don't follow -- if "this impl" refers to `impl forall [T:! Second] T as First`, I don't see why "does `T` impl `Second`" would look at it. If it refers to `impl forall [T:! type] T as Second`, I don't see why it would lead to a recursive query.

Reply:
Impl lookup just does a for loop over all the impls. This impl comes first, so it looks at it, but we have an early out in impl_lookup to avoid going further when it sees `as First` does not match the query for `Second`.

This also made an infinite loop until a week or so ago, but I got that working. This is the simplest case and in more complex cases (more complex queries, or generic interfaces like below) we end up trying to look at the impl, in which case we deduce arguments and loop back to looking at impls again and find ourselves at the same impl.

Reply:
OK, so it's not that this _requires_ a recursive query in deduction, it's that naively this _could_ be implemented in terms of a recursive query, and this test is validating that we in fact avoid the recursive query?

Reply:
Yes!

Reply:
OK, I've added a suggestion to try to make all this clearer.